utilities/transactions/lock/range/range_tree/range_tree_lock_manager.cc

```diff
@@ -0,0 +1,572 @@
+  }
+}
+
+RangeTreeLockManager::~RangeTreeLockManager() {
+  autovector<void*> local_caches;
+  ltree_lookup_cache_->Scrape(&local_caches, nullptr);
+  for (auto cache : local_caches) {
+    delete static_cast<LockTreeMap*>(cache);
+  }
+
+  for (auto it : ltree_map_) {
+    ltm_.release_lt(it.second);
+  }
+  ltm_.destroy();
+}
+
+RangeLockManagerHandle::Counters RangeTreeLockManager::GetStatus() {
+  LTM_STATUS_S ltm_status_test;
+  ltm_.get_status(&ltm_status_test);
+  Counters res;
+
+  // Searching status variable by its string name is how Toku's unit tests
+  // do it (why didn't they make LTM_ESCALATION_COUNT constant visible?)
+  // lookup keyname in status
+  for (int i = 0; i < LTM_STATUS_S::LTM_STATUS_NUM_ROWS; i++) {
+    TOKU_ENGINE_STATUS_ROW status = &ltm_status_test.status[i];
+    if (strcmp(status->keyname, "LTM_ESCALATION_COUNT") == 0) {
+      res.escalation_count = status->value.num;
+      continue;
+    }
+    if (strcmp(status->keyname, "LTM_SIZE_CURRENT") == 0) {
+      res.current_lock_memory = status->value.num;
+    }
+  }
+  return res;
+}
+
+void RangeTreeLockManager::AddColumnFamily(const ColumnFamilyHandle* cfh) {
+  uint32_t column_family_id = cfh->GetID();
+
+  InstrumentedMutexLock l(&ltree_map_mutex_);
+  if (ltree_map_.find(column_family_id) == ltree_map_.end()) {
+    DICTIONARY_ID dict_id = {.dictid = column_family_id};
+    toku::comparator cmp;
+    cmp.create(compare_dbt_endpoints, (void*)cfh->GetComparator());
+    toku::locktree* ltree = ltm_.get_lt(dict_id, cmp,
+                                        /* on_create_extra*/ nullptr);
+    // This is ok to because get_lt has copied the comparator:
+    cmp.destroy();
+    ltree_map_.emplace(column_family_id, ltree);
+  } else {
+    // column_family already exists in lock map
+    // assert(false);
+  }
+}
+
+void RangeTreeLockManager::RemoveColumnFamily(const ColumnFamilyHandle* cfh) {
+  uint32_t column_family_id = cfh->GetID();
+  // Remove lock_map for this column family.  Since the lock map is stored
+  // as a shared ptr, concurrent transactions can still keep using it
+  // until they release their references to it.
+
+  // TODO^ if one can drop a column family while a transaction is holding a
+  // lock in it, is column family's comparator guaranteed to survive until
+  // all locks are released? (we depend on this).
+  {
+    InstrumentedMutexLock l(&ltree_map_mutex_);
+
+    auto lock_maps_iter = ltree_map_.find(column_family_id);
+    assert(lock_maps_iter != ltree_map_.end());
+
+    ltm_.release_lt(lock_maps_iter->second);
+
+    ltree_map_.erase(lock_maps_iter);
+  }  // lock_map_mutex_
+
+  // TODO: why do we delete first and clear the caches second? Shouldn't this be
+  // done in the reverse order? (if we do it in the reverse order, how will we
+  // prevent somebody from re-populating the cache?)
+
+  // Clear all thread-local caches. We collect a vector of caches but we dont
+  // really need them.
+  autovector<void*> local_caches;
```

Comment:
what if a lock tree is already released by `ltm_.release_lt`, but not cleared from thread local cache yet, and someone accesses it from the local cache?

Reply:
Ok, there was a bug there. Now, I've made it work like point-based lock manager does